### PR TITLE
Logging

### LINKED ISSUE #7: configurable logging
* [ ] let the logging be configurable
* [ ] every python file has a logger on their own

### PR DESCRIPTION
**Pull request by TomasTomecek**
Only https://github.com/TomasTomecek/conu/commit/ef7b02838148b9509ebfe270d3082f6406d15089 is part of this PR. Will rebase once #21 is merged. Commit message:

```
better logging

* per-module logger
* configurable library-specific logger via Backend interface
* unify logging messages (lower-case start)

Fixes #7

Built on top of #21, will rebase once merged
```

### CODE REVIEW

**Comment by phracek** [comment]
One nit. Why in Pull Request Logging is implemented http client for container?

File: conu/apidefs/backend.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+"""
+Definition for a backend class
+"""
+
+import logging
+
+from .container import Container
+from .image import Image
```
Comment (phracek) [inline conu/apidefs/backend.py]: Is there any reason, why you use relative path? I guess, `from conu.apidefs.container` is better from readability point of view.

File: conu/apidefs/container.py [hunk 2]
```diff
@@ -2,9 +2,11 @@
 Abstract class definitions for containers.
 """
 
-# FIXME: TBD
 from .image import Image
```
Comment (phracek) [inline conu/apidefs/container.py]: The same here with relative path.

File: conu/apidefs/backend.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+"""
+Definition for a backend class
+"""
+
+import logging
+
+from .container import Container
```
Comment (phracek) [inline conu/apidefs/backend.py]: This should be `DockerContainer` and not `Container`, right?

File: conu/backend/docker/__init__.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# -*- coding: utf-8 -*-
+
+from .backend import DockerBackend
+from .container import DockerContainer
+from .image import DockerImage
```
Comment (phracek) [inline conu/backend/docker/__init__.py]: Why relative paths?

File: conu/apidefs/backend.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+from .image import Image
+
+
+__version__ = "0.0.1-alpha"
+
+
+def set_logging(level=logging.INFO):
```
Comment (phracek) [inline conu/apidefs/backend.py]: I would prefer to create a class with several static classes.

File: conu/apidefs/backend.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    :param level: int, see logging.{DEBUG,INFO,ERROR,...}
+    :return: logger instance
+    """
+    logger = logging.getLogger("conu")
+    # do we want to propagate to root logger?
+    # logger.propagate = False
+    logger.setLevel(level)
+
+    # logging to std.err
+    handler = logging.StreamHandler()
```
Comment (phracek) [inline conu/apidefs/backend.py]: What about if I would like to use FileHandler? I would prefer to create a function which will consume only handler.
Does it make sense to have FileHandler, at all?

File: tests/integration/test_docker.py [hunk 7]
```diff
@@ -70,19 +69,34 @@ def test_docker():
...
+    assert c.check_running()
+    r = c.http_request(port="8000")
+    assert "<!DOCTYPE HTML PUBLIC" in r.content
+    assert r.ok
+    r2 = c.http_request(path="/etc", port="8000")
+    assert "<!DOCTYPE HTML PUBLIC" in r2.content
+    assert "passwd" in r2.content
+    assert r2.ok
+
+
```
Comment (phracek) [inline tests/integration/test_docker.py]: This should be part of another PR and not related to Logging.

File: conu/apidefs/backend.py (hunk 3 above)
Comment (TomasTomecek) [reply conu/apidefs/backend.py]: no, this is the common interface which is shared for all the implementations

File: conu/apidefs/backend.py (hunk 1 above)
Comment (TomasTomecek) [reply conu/apidefs/backend.py]: I guess it's a matter of personal taste, I don't mind changing all to absolute -- my version is shorter and in my opinion obvious; what do others think? @dhodovsk @jscotka @bkabrda

File: conu/apidefs/backend.py (hunk 5 above)
Comment (TomasTomecek) [reply conu/apidefs/backend.py]: I don't understand, can you provide an example?

File: conu/apidefs/backend.py (hunk 6 above)
Comment (TomasTomecek) [reply conu/apidefs/backend.py]: we can parametrize the function: this is a good suggestion, not sure if anyone is interested in logging to a file

File: tests/integration/test_docker.py (hunk 7 above)
Comment (TomasTomecek) [reply tests/integration/test_docker.py]: This is part of another PR, see the description of this PR.

File: conu/apidefs/backend.py (hunk 5 above)
Comment (phracek) [reply conu/apidefs/backend.py]: Like this https://github.com/rebase-helper/rebase-helper/blob/master/rebasehelper/logger.py

File: conu/apidefs/backend.py (hunk 6 above)
Comment (phracek) [reply conu/apidefs/backend.py]: Good example is here: https://github.com/rebase-helper/rebase-helper/blob/master/rebasehelper/logger.py

File: conu/apidefs/backend.py (hunk 5 above)
Comment (TomasTomecek) [reply conu/apidefs/backend.py]: I can see that for rebase helper it works nicely. For conu, it's just too complicated -- I doubt this project will ever have more than one (at most two) loggers -- therefore I think that having a single function to initiate logging is simpler for current use case. Once we are in a need to have multiple loggers using different configuration, we can revisit.

File: conu/apidefs/backend.py (hunk 6 above)
Comment (TomasTomecek) [reply conu/apidefs/backend.py]: see my comment above

**Comment by TomasTomecek** [comment]
I believe I addressed all the comments, please take a look.

**Comment by phracek** [comment]
LGTM

**Comment by TomasTomecek** [comment]
\o/

**Comment by psklenar** [comment]
## jenkins/superCI comment
* Status: failure, return code = __1__ 
* Result url: __[https://da.gd/1UVN](https://da.gd/1UVN)__
* Comment: __This is testing CI__
